File: internal/screenbench/scenario/scenario.go
// Package scenario loads recorded bake-off scenarios from disk.
//
// A scenario is a directory laid out as:
//
//	scenarios/<harness>/<name>/
//	    bytes.raw          required: raw PTY byte stream captured from the harness
//	    meta.json          required: harness, recorded_at, terminal dims, binary version
//	    expected.txt       optional: ground-truth final assistant text
//	    transcript.jsonl   optional: copy of the harness's own session log
//
// The bench replays bytes.raw through each emulator and compares the
// resulting screen snapshot against expected.txt.
package scenario

import (
	"encoding/json"
	"errors"
	"fmt"
	"io/fs"
	"os"
	"path/filepath"
	"sort"
	"time"
)

// Meta is the parsed contents of meta.json.
type Meta struct {
	Harness       string    `json:"harness"`        // e.g. "codex", "claude-code"
	BinaryVersion string    `json:"binary_version"` // e.g. "codex 0.42.1"
	RecordedAt    time.Time `json:"recorded_at"`
	Cols          int       `json:"cols"`
	Rows          int       `json:"rows"`
	Notes         string    `json:"notes,omitempty"`
}

// Scenario is one loaded corpus entry.
type Scenario struct {
	Name     string
	Path     string
	Meta     Meta
	Bytes    []byte
	Expected string // contents of expected.txt; empty if missing
}

// Load loads a single scenario directory.
func Load(dir string) (*Scenario, error) {
	metaBytes, err := os.ReadFile(filepath.Join(dir, "meta.json"))
	if err != nil {
		return nil, fmt.Errorf("read meta.json: %w", err)
	}
	var meta Meta
	if err := json.Unmarshal(metaBytes, &meta); err != nil {
		return nil, fmt.Errorf("parse meta.json: %w", err)
	}
	if meta.Cols == 0 {
		meta.Cols = 120
	}
	if meta.Rows == 0 {
		meta.Rows = 40
	}
	rawBytes, err := os.ReadFile(filepath.Join(dir, "bytes.raw"))
	if err != nil {
		return nil, fmt.Errorf("read bytes.raw: %w", err)
	}
	expected, err := os.ReadFile(filepath.Join(dir, "expected.txt"))
	if err != nil && !errors.Is(err, fs.ErrNotExist) {
		return nil, fmt.Errorf("read expected.txt: %w", err)
	}
	return &Scenario{
		Name:     filepath.Base(dir),
		Path:     dir,
		Meta:     meta,
		Bytes:    rawBytes,
		Expected: string(expected),
	}, nil
}

// Discover walks root and returns every scenario directory found.
// A directory qualifies as a scenario iff it contains meta.json.
func Discover(root string) ([]*Scenario, error) {
	var out []*Scenario
	err := filepath.WalkDir(root, func(path string, d fs.DirEntry, err error) error {
		if err != nil {
			if errors.Is(err, fs.ErrNotExist) {
				return nil
			}
			return err
		}
		if !d.IsDir() {
			return nil
		}
		if _, err := os.Stat(filepath.Join(path, "meta.json")); err != nil {
			return nil
		}
		s, err := Load(path)
		if err != nil {
			return fmt.Errorf("load %s: %w", path, err)
		}
		out = append(out, s)
		return nil
	})
	if err != nil {
		return nil, err
	}
	sort.Slice(out, func(i, j int) bool {
		if out[i].Meta.Harness != out[j].Meta.Harness {
			return out[i].Meta.Harness < out[j].Meta.Harness
		}
		return out[i].Name < out[j].Name
	})
	return out, nil
}

// WriteMeta is a convenience used by the recorder to emit meta.json.
func WriteMeta(dir string, m Meta) error {
	if err := os.MkdirAll(dir, 0o755); err != nil {
		return err
	}
	b, err := json.MarshalIndent(m, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(filepath.Join(dir, "meta.json"), b, 0o644)
}

File: test/corpus/tools/screenbench/scenario.ts
// Port of internal/screenbench/scenario/scenario.go: loads recorded bake-off
// scenarios from disk.
//
// A scenario is a directory laid out as:
//
//   scenarios/<harness>/<name>/
//       bytes.raw          required: raw PTY byte stream captured from the harness
//       meta.json          required: harness, recorded_at, terminal dims, binary version
//       expected.txt       optional: ground-truth final assistant text
//       transcript.jsonl   optional: copy of the harness's own session log
//
// The bench replays bytes.raw through each emulator and compares the
// resulting screen snapshot against expected.txt.

import { promises as fs } from "node:fs"
import * as path from "node:path"

/** Parsed contents of meta.json. */
export interface Meta {
  harness: string // e.g. "codex", "claude-code"
  binaryVersion: string // e.g. "codex 0.42.1"
  recordedAt: string
  cols: number
  rows: number
  notes?: string
}

/** On-disk JSON shape of meta.json (Go's `encoding/json` field names). */
interface RawMeta {
  harness: string
  binary_version: string
  recorded_at: string
  cols?: number
  rows?: number
  notes?: string
}

/** One loaded corpus entry. */
export interface Scenario {
  name: string
  path: string
  meta: Meta
  bytes: Buffer
  expected: string // contents of expected.txt; empty if missing
}

function wrapError(label: string, err: unknown): Error {
  const message = err instanceof Error ? err.message : String(err)
  return new Error(`${label}: ${message}`, { cause: err })
}

function isNotFound(err: unknown): boolean {
  return typeof err === "object" && err !== null && (err as NodeJS.ErrnoException).code === "ENOENT"
}

/** Loads a single scenario directory. */
export async function load(dir: string): Promise<Scenario> {
  let metaBytes: Buffer
  try {
    metaBytes = await fs.readFile(path.join(dir, "meta.json"))
  } catch (err) {
    throw wrapError("read meta.json", err)
  }

  let raw: RawMeta
  try {
    raw = JSON.parse(metaBytes.toString("utf8"))
  } catch (err) {
    throw wrapError("parse meta.json", err)
  }

  const meta: Meta = {
    harness: raw.harness,
    binaryVersion: raw.binary_version,
    recordedAt: raw.recorded_at,
    cols: raw.cols || 120,
    rows: raw.rows || 40,
    notes: raw.notes,
  }

  let rawBytes: Buffer
  try {
    rawBytes = await fs.readFile(path.join(dir, "bytes.raw"))
  } catch (err) {
    throw wrapError("read bytes.raw", err)
  }

  let expected = ""
  try {
    expected = (await fs.readFile(path.join(dir, "expected.txt"))).toString("utf8")
  } catch (err) {
    if (!isNotFound(err)) {
      throw wrapError("read expected.txt", err)
    }
  }

  return {
    name: path.basename(dir),
    path: dir,
    meta,
    bytes: rawBytes,
    expected,
  }
}

async function walk(dir: string, acc: string[]): Promise<void> {
  let entries: import("node:fs").Dirent[]
  try {
    entries = await fs.readdir(dir, { withFileTypes: true })
  } catch (err) {
    if (isNotFound(err)) {
      return
    }
    throw err
  }
  acc.push(dir)
  for (const entry of entries) {
    if (entry.isDirectory()) {
      await walk(path.join(dir, entry.name), acc)
    }
  }
}

/**
 * Walks root and returns every scenario directory found. A directory
 * qualifies as a scenario iff it directly contains a meta.json file.
 */
export async function discover(root: string): Promise<Scenario[]> {
  const dirs: string[] = []
  await walk(root, dirs)

  const out: Scenario[] = []
  for (const dir of dirs) {
    try {
      await fs.stat(path.join(dir, "meta.json"))
    } catch {
      continue
    }
    try {
      out.push(await load(dir))
    } catch (err) {
      throw wrapError(`load ${dir}`, err)
    }
  }

  out.sort((a, b) => {
    if (a.meta.harness !== b.meta.harness) {
      return a.meta.harness < b.meta.harness ? -1 : 1
    }
    if (a.name !== b.name) {
      return a.name < b.name ? -1 : 1
    }
    return 0
  })

  return out
}

/** Convenience used by the recorder to emit meta.json. */
export async function writeMeta(dir: string, meta: Meta): Promise<void> {
  await fs.mkdir(dir, { recursive: true })
  const raw: RawMeta = {
    harness: meta.harness,
    binary_version: meta.binaryVersion,
    recorded_at: meta.recordedAt,
    cols: meta.cols,
    rows: meta.rows,
  }
  if (meta.notes) {
    raw.notes = meta.notes
  }
  const json = JSON.stringify(raw, null, 2)
  await fs.writeFile(path.join(dir, "meta.json"), json)
}
